sway (1.6) / wlroots (0.13.0) broke screen sharing and gnome3 pinentry popups

**Describe the bug**
The recent update of sway and wlroots seems to have broken screen sharing as well as gnome3 pinentry popups.

**To Reproduce**
Steps to reproduce the behavior:

I've built a system on 311ceed827f531f88f46222920cd1ebb2c101f73 (current nixos-unstable, there neither of the mentioned things works).

But if I build a system on 04a2b269d8921505a2969fc9ec25c1f517f2b307 (previous unstable, there both of the mentioned things works.).

I've then checked out 04a2b269d8921505a2969fc9ec25c1f517f2b307 and copied the folders `pkgs/development/libraries/wlroots/` and `pkgs/applications/window-managers/sway/` from master into my checkout and built a system based on the old unstable but with only those two packages updated. Then it's still broken. So it's at least isolated to those two packages.

**Expected behavior**
I expect screen sharing to work.

I also expect to be able to unlock my security key from places where the ncurses interface doesn't work (like within Emacs).

**Additional context**
Here's my complete sway setup: https://github.com/etu/nixconfig/blob/main/modules/my-sway.nix#L465-L516

**Notify maintainers**
cc @primeos @Synthetica9 @Ma27 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.11.9, NixOS, 21.05.git.04a2b269d89M (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `""`
 - nixpkgs: `/etc/nixos/nix/nixos-unstable`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: sway wlroots
# a list of nixos modules affected by the problem
module:
```

`xdg-desktop-portal-wlr.service` fails to start for me on nixos-unstable; presumably this is why screen sharing doesn't work. This appears in `journalctl`:

```
Apr 14 20:02:20 girafarig systemd[1244]: Starting Portal service (wlroots implementation)...
Apr 14 20:02:20 girafarig xdg-desktop-portal-wlr[1363]: 2021/04/14 20:02:20 [ERROR] - wayland: failed to connect to display
Apr 14 20:02:20 girafarig systemd[1244]: xdg-desktop-portal-wlr.service: Main process exited, code=exited, status=1/FAILURE
Apr 14 20:02:20 girafarig systemd[1244]: xdg-desktop-portal-wlr.service: Failed with result 'exit-code'.
Apr 14 20:02:20 girafarig systemd[1244]: Failed to start Portal service (wlroots implementation).
```

--- a/nixos/modules/programs/sway.nix
+++ b/nixos/modules/programs/sway.nix
@@ -31,6 +31,7 @@ let
     extraOptions = cfg.extraOptions;
     withBaseWrapper = cfg.wrapperFeatures.base;
     withGtkWrapper = cfg.wrapperFeatures.gtk;
+    isNixOS = true;
   };
 in {
   options.programs.sway = {
@@ -120,8 +121,11 @@ in {
       systemPackages = [ swayPackage ] ++ cfg.extraPackages;
       etc = {
         "sway/config".source = mkOptionDefault "${swayPackage}/etc/sway/config";
-        #"sway/security.d".source = mkOptionDefault "${swayPackage}/etc/sway/security.d/";
-        #"sway/config.d".source = mkOptionDefault "${swayPackage}/etc/sway/config.d/";
+        "sway/config.d/nixos.conf".source = pkgs.writeText "nixos.conf" ''
+          # Import the most important environment variables into the D-Bus and systemd
+          # user environments (e.g. required for screen sharing and Pinentry prompts):
+          exec dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY SWAYSOCK
+        '';
       };
     };
     security.pam.services.swaylock = {};

--- a/pkgs/applications/window-managers/sway/default.nix
+++ b/pkgs/applications/window-managers/sway/default.nix
@@ -4,6 +4,8 @@
 , pango, cairo, libinput, libcap, pam, gdk-pixbuf, librsvg
 , wlroots, wayland-protocols, libdrm
 , nixosTests
+# Used by the NixOS module:
+, isNixOS ? false
 }:
 
 stdenv.mkDerivation rec {
@@ -26,6 +28,10 @@ stdenv.mkDerivation rec {
       inherit swaybg;
     })
   ];
+
+  postPatch = lib.optionalString isNixOS ''
+    echo -e '\ninclude /etc/sway/config.d/*' >> config.in
+  '';
 
   nativeBuildInputs = [
     meson ninja pkg-config wayland scdoc

--- a/pkgs/applications/window-managers/sway/wrapper.nix
+++ b/pkgs/applications/window-managers/sway/wrapper.nix
@@ -4,6 +4,8 @@
 , withBaseWrapper ? true, extraSessionCommands ? "", dbus
 , withGtkWrapper ? false, wrapGAppsHook, gdk-pixbuf, glib, gtk3
 , extraOptions ? [] # E.g.: [ "--verbose" ]
+# Used by the NixOS module:
+, isNixOS ? false
 }:
 
 assert extraSessionCommands != "" -> withBaseWrapper;
@@ -11,6 +13,7 @@ assert extraSessionCommands != "" -> withBaseWrapper;
 with lib;
 
 let
+  sway = sway-unwrapped.override { inherit isNixOS; };
   baseWrapper = writeShellScriptBin "sway" ''
      set -o errexit
      if [ ! "$_SWAY_WRAPPER_ALREADY_EXECUTED" ]; then
@@ -19,16 +22,16 @@ let
      fi
      if [ "$DBUS_SESSION_BUS_ADDRESS" ]; then
        export DBUS_SESSION_BUS_ADDRESS
-       exec ${sway-unwrapped}/bin/sway "$@"
+       exec ${sway}/bin/sway "$@"
      else
-       exec ${dbus}/bin/dbus-run-session ${sway-unwrapped}/bin/sway "$@"
+       exec ${dbus}/bin/dbus-run-session ${sway}/bin/sway "$@"
      fi
    '';
 in symlinkJoin {
-  name = "sway-${sway-unwrapped.version}";
+  name = "sway-${sway.version}";
 
   paths = (optional withBaseWrapper baseWrapper)
-    ++ [ sway-unwrapped ];
+    ++ [ sway ];
 
   nativeBuildInputs = [ makeWrapper ]
     ++ (optional withGtkWrapper wrapGAppsHook);
@@ -48,5 +51,5 @@ in symlinkJoin {
 
   passthru.providedSessions = [ "sway" ];
 
-  inherit (sway-unwrapped) meta;
+  inherit (sway) meta;
 }

